pkg/hubble/metrics/api/context.go

```diff
@@ -129,30 +159,123 @@ func parseContext(s string) (cs ContextIdentifierList, err error) {
 // relevant options
 func ParseContextOptions(options Options) (*ContextOptions, error) {
 	o := &ContextOptions{}
+	var err error
 	for key, value := range options {
 		switch strings.ToLower(key) {
 		case "destinationcontext":
-			c, err := parseContext(value)
+			o.Destination, err = parseContext(value)
 			if err != nil {
 				return nil, err
 			}
-			o.Destination = c
 		case "sourcecontext":
-			c, err := parseContext(value)
+			o.Source, err = parseContext(value)
+			if err != nil {
+				return nil, err
+			}
+		case "labelscontext":
+			o.Labels, err = parseLabels(value)
 			if err != nil {
 				return nil, err
 			}
-			o.Source = c
 		}
 	}
 
 	return o, nil
 }
 
+type labelsSet map[string]struct{}
+
+func newLabelsSet(labels []string) labelsSet {
+	m := make(map[string]struct{}, len(labels))
+	for _, label := range labels {
+		m[label] = struct{}{}
+	}
+	return labelsSet(m)
+}
+
+func (ls labelsSet) HasLabel(label string) bool {
+	_, exists := ls[label]
+	return exists
+}
+
+func labelsContext(wantedLabels labelsSet, flow *pb.Flow) (outputLabels []string) {
+	// Iterate over contextLabelsList so that the label order is stable,
+	// otherwise GetLabelNames and GetLabelValues might be mismatched
+	for _, label := range contextLabelsList {
+		if wantedLabels.HasLabel(label) {
+			var labelValue string
+			switch label {
+			case "source_pod":
+				if flow.GetSource() != nil {
+					labelValue = flow.GetSource().PodName
+				}
+			case "source_namespace":
+				if flow.GetSource() != nil {
+					labelValue = flow.GetSource().Namespace
+				}
+			case "source_workload":
+				if flow.GetSource() != nil {
+					if workloads := flow.GetSource().GetWorkloads(); len(workloads) != 0 {
+						labelValue = workloads[0].Name
+					}
+				}
+			case "source_app":
+				if flow.GetSource() != nil {
+					labelValue = getK8sAppFromLabels(flow.GetSource().GetLabels())
+				}
+			case "destination_pod":
+				if flow.GetDestination() != nil {
+					labelValue = flow.GetDestination().PodName
+				}
+			case "destination_namespace":
+				if flow.GetDestination() != nil {
+					labelValue = flow.GetDestination().Namespace
+				}
+			case "destination_workload":
+				if flow.GetDestination() != nil {
+					if workloads := flow.GetDestination().GetWorkloads(); len(workloads) != 0 {
+						labelValue = workloads[0].Name
+					}
+				}
+			case "destination_app":
+				if flow.GetDestination() != nil {
+					labelValue = getK8sAppFromLabels(flow.GetDestination().GetLabels())
+				}
+			case "reporter":
+				switch flow.GetTrafficDirection() {
+				case pb.TrafficDirection_EGRESS:
+					labelValue = "client"
+				case pb.TrafficDirection_INGRESS:
+					labelValue = "server"
+				default:
+					labelValue = "unknown"
+				}
+			default:
+				// Label is in contextLabelsList but isn't handled the switch
+				// statement. Programmer error.
+				panic(fmt.Sprintf("label %s not mapped in labelsContext, please update labelsContext", label))
```

Comment:
Hmm I see your point but, these functions don't return errors so I wasn't sure of a better better approach without refactoring GetLabelValues to return an error, which means all plugin handlers will have to handle that. The other option is that we turn these into some kind of method and provide a logger to the object so it can log the error.

In any case, if we don't produce the right number of output labels, the prometheus instrumentation will throw an error since they labels values list won't match the length of the labels list.

I guess if prometheus will error, there's no point in the logger approach, so I think returning the error would be best in this case, even though it should never be hit unless we have a bug.

Another thought is we could add a `recover()` around the hubble server start and catch hubble panics if we don't want hubble bugs to crash cilium.

Reply:
Actually, even if we return the error, metrics plugin `ProcessFlow` handlers don't return errors either, and none of the handlers get plumbed a logger either.

Reply:
Ok, added a new commit to replace the panic. It involved some refactoring so I opted to separate that commit out from the rest.